ui/src/minard/utils/bin.ts

```diff
@@ -0,0 +1,161 @@
+        10 |   20 |    0 |    4 |  "a" |   1
+        10 |   20 |    4 |    6 |  "b" |   1
+
+  If `binCount` is not provided, a default value will be provided using
+  [Sturges' formula][0].
+
+  [0]: https://en.wikipedia.org/wiki/Histogram#Sturges'_formula
+*/
+export const bin = (
+  table: Table,
+  xColKey: string,
+  groupColKeys: string[] = [],
+  binCount: number,
+  position: HistogramPosition
+) => {
+  const xCol = table.columns[xColKey]
+  const xColType = table.columnTypes[xColKey]
+
+  assert(`could not find column "${xColKey}"`, !!xCol)
+  assert(
+    `unsupported value column type "${xColType}"`,
+    xColType === ColumnType.Numeric || xColType === ColumnType.Temporal
+  )
+
+  const bins = createBins(xCol, binCount)
+  const groups = {}
+
+  // Count x values by bin and group
+  for (let i = 0; i < xCol.length; i++) {
+    const xDatum = xCol[i]
+    const groupData = getGroupData(table, groupColKeys, i)
+    const groupKey = getGroupKey(Object.values(groupData))
+    const bin = bins.find(
+      (b, i) => (xDatum < b.max && xDatum >= b.min) || i === bins.length - 1
+    )
+
+    groups[groupKey] = groupData
+
+    if (!bin.values[groupKey]) {
+      bin.values[groupKey] = 1
+    } else {
+      bin.values[groupKey] += 1
+    }
+  }
+
+  // Then build up a tabular representation of each of these bins by group
+  const statTable = {
+    columns: {xMin: [], xMax: [], yMin: [], yMax: []},
+    columnTypes: {
+      xMin: xColType,
+      xMax: xColType,
+      yMin: ColumnType.Numeric,
+      yMax: ColumnType.Numeric,
+    },
+  }
+
+  for (const key of groupColKeys) {
+    statTable.columns[key] = []
+    statTable.columnTypes[key] = table.columnTypes[key]
+  }
+
+  const orderedGroupKeys = Object.keys(groups)
+
+  for (let i = 0; i < orderedGroupKeys.length; i++) {
+    const groupKey = orderedGroupKeys[i]
+
+    for (const bin of bins) {
+      let yMin = 0
+
+      if (position === HistogramPosition.Stacked) {
+        yMin = orderedGroupKeys
+          .slice(0, i)
+          .reduce((sum, k) => sum + (bin.values[k] || 0), 0)
+      }
+
+      statTable.columns.xMin.push(bin.min)
+      statTable.columns.xMax.push(bin.max)
+      statTable.columns.yMin.push(yMin)
+      statTable.columns.yMax.push(yMin + (bin.values[groupKey] || 0))
+
+      for (const [k, v] of Object.entries(groups[groupKey])) {
+        statTable.columns[k].push(v)
+      }
+    }
+  }
+
+  const mappings: any = {
+    xMin: 'xMin',
+    xMax: 'xMax',
+    yMin: 'yMin',
+    yMax: 'yMax',
+    fill: groupColKeys,
+  }
+
+  return [statTable, mappings]
+}
+
+const createBins = (
+  col: number[],
+  binCount: number
+): Array<{max: number; min: number; values: {}}> => {
+  if (!binCount) {
+    binCount = thresholdSturges(col)
+  }
+
+  const domain = extent(col)
+  const d0 = domain[0]
+
+  let d1 = domain[1]
+
+  if (d0 === d1) {
+    d1 = d0 + 1
+  }
+
+  const bins = range(d0, d1, (d1 - d0) / binCount).map(min => ({
```

Comment:
This line is very concise, but also hard to parse visually. If you're thinking of this as a separate library and util functions I think it would be better to err on the side of verbosity instead. This would make it easy for the next person to look at the code to be able to reason about it. This is my general comment on this file and others in minard.